### PR TITLE
issue: 1081001 Add netmap API VMA wrapper

### PR DESCRIPTION
**Pull request by serglelv**
The netmap_vma API is an adaptation layer from netmap API
to VMA striding RQ API. Only RX is supported.

Signed-off-by: Sergey Lykov <sergeyly@mellanox.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1608/ for details (Mellanox internal link).

**Comment by serglelv** [comment]
Please review.
@rafiw @ilansmith @liranoz12

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 1]
```diff
@@ -0,0 +1,795 @@
...
+			+ (unsigned long long int) tv.tv_usec);
+}
+
+static int CreateRingProfile(bool CommonFdPerRing, int RingProfile, int user_id, int RxSocket )
+{
+	vma_ring_alloc_logic_attr profile;
+	profile.engress = 0;
+	profile.ingress = 1;
+	profile.ring_profile_key = RingProfile;
+	if (CommonFdPerRing ) {
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: It seems yes.

File: tools/media/netmap-vma/apps/makefile [hunk 2]
```diff
@@ -0,0 +1,28 @@
+
+NETMAP_SRC_DIR = /path_to_the_netmap_sources
```
Comment (serglelv) [inline tools/media/netmap-vma/apps/makefile]: The command make NETMAP_SRC_DIR=/path BITSTREAM_DIR=/path can be used

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 3]
```diff
@@ -0,0 +1,795 @@
...
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+
+#include <unistd.h>
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Code cleanup

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 4]
```diff
@@ -0,0 +1,795 @@
...
+#include <dlfcn.h>
+
+#include "vma/util/vtypes.h"
+#include "vma_extra.h"
+
+#ifdef HAVE_MP_RQ
+#define HAVE_MP_RQ_NETMAP
+#define NETMAP_WITH_LIBS
+#include <net/netmap_user.h>
+#endif
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Agree. It will be better.

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 5]
```diff
@@ -0,0 +1,795 @@
...
+#define	MAX_RINGS	1000
+#define	PRINT_PERIOD_SEC	5
+#define	PRINT_PERIOD		1000000 * PRINT_PERIOD_SEC
+#define	MAX_SOCKETS_PER_RING	4096
+#define	STRIDE_SIZE	2048
+
+class RXSock;
+class CommonCyclicRing;
+
+typedef void (*validatePackets)(uint8_t*, size_t, CommonCyclicRing*);
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Code cleanup

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 6]
```diff
@@ -0,0 +1,795 @@
...
+		pRings[ring]->completion.packets = 0;
+		if (pRings[ring]->vma_api->vma_cyclic_buffer_read(pRings[ring]->ring_fd, &pRings[ring]->completion, pRings[ring]->min_s, pRings[ring]->max_s, pRings[ring]->flags) < 0) {
+			return -1;
+		}
+		if (pRings[ring]->completion.packets) {
+			pRings[ring]->is_readable = true;
+			return 1;
+		}
+	}
+	usleep(pRings[ring]->sleep_time);
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Currently disabled.

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 7]
```diff
@@ -0,0 +1,795 @@
...
+		opts++;
+	}
+
+	std::string nmring;
+	nmring.append(nm_ring, nm_ring_len);
+
+	string nmport;
+	if ( nm_port == NULL) {
+		string default_port;
+		cout << default_port << DEFAULT_PORT << endl;
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Agree. Updated.

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 8]
```diff
@@ -0,0 +1,795 @@
...
+			unsigned short hash = hashIpPort2(**it);
+			//printf("hash value is %d\n",hash);
+			if ( NULL != pRings[i]->hashedSock[hash] ) {
+				printf ("Collision, reshuffle your ip addresses \n");
+				return NULL;
+			}
+			pRings[i]->hashedSock[hash] = pSock;
+			inet_ntop(AF_INET, &((*it)->sin_addr), pSock->ipAddress, INET_ADDRSTRLEN);
+			pRings[i]->sock_vect.push_back(pSock);
+			pRings[i]->min_s = 1000;
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Agree. Updated

File: tools/media/netmap-vma/apps/readme.txt [hunk 9]
```diff
@@ -0,0 +1,51 @@
+
```
Comment (serglelv) [inline tools/media/netmap-vma/apps/readme.txt]: Done

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 10]
```diff
@@ -0,0 +1,795 @@
...
+	} else {
+		cnf_if.push_back(opts_line);
+	}
+
+	while (!cnf_if.empty()) {
+		line = cnf_if.back();
+		cnf_if.pop_back();
+		std::istringstream iss(line);
+		struct flow_param flow;
+		if (iss >> ip >> port) {
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Updated

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 11]
```diff
@@ -0,0 +1,795 @@
...
+	int port;
+	int ring_id;
+	int lineNum = 0;
+	int sock_num = 0, socketRead = 0;
+	char HashColision[MAX_RINGS][MAX_SOCKETS_PER_RING] = {0};
+	int uniqueRings;
+	int hash_colision_cnt = 0;
+	char *cnfif_file = new char[IFNAMSIZ+4];
+
+	strcpy(cnfif_file,ifname);
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Agree. Updated

File: tools/media/netmap-vma/netmap_vma.cpp [hunk 12]
```diff
@@ -0,0 +1,795 @@
...
+	char ifname[IFNAMSIZ];
+	char nm_ring_val[10];
+	u_int namelen;
+	struct nm_desc *d = NULL;
+	int nm_ring_len = 0;
+	int nm_ring_set = 0;
+	struct ifaddrs *ifaddr, *ifa;
+	char host[NI_MAXHOST];
+
+	d = (struct nm_desc *)calloc(1, sizeof(*d));
```
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: Updated

File: tools/media/netmap-vma/netmap_vma.cpp (hunk 1 above)
Comment (serglelv) [inline tools/media/netmap-vma/netmap_vma.cpp]: It seems yes.

**Comment by serglelv** [comment]
Updated. Please review.
@rafiw

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1614/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1618/ for details (Mellanox internal link).

File: tools/media/netmap-vma/apps/makefile (hunk 2 above)
Comment (rafiw) [inline tools/media/netmap-vma/apps/makefile]: good to know

File: tools/media/netmap-vma/netmap_vma.cpp (hunk 1 above)
Comment (rafiw) [inline tools/media/netmap-vma/netmap_vma.cpp]: ok

**Comment by serglelv** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1622/ for details (Mellanox internal link).

**Comment by liranoz12** [comment]
Merged #449.

**Pull request closed without merging**